components/eamxx/src/diagnostics/tests/horiz_avg_test.cpp

```diff
@@ -0,0 +1,226 @@
+        local_wavg += (area_v[icol] / atot) * qc1_v[icol];
+      },
+      wavg);
+  Kokkos::deep_copy(diag0_v, wavg);
+
+  auto diag1_v = diag1_f.get_view<Real>();
+
+  bool result = false;
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<>(0, 1),
+      KOKKOS_LAMBDA(const int, bool &local_result) {
+        local_result = (diag1_v() == diag0_v());
+      },
+      result);
+  REQUIRE(result);
+
+  // Try known cases
+  // Set qc1_v to 1.0 to get weighted average of 1.0
+  Kokkos::deep_copy(qc1_v, 1.0);
+  Kokkos::deep_copy(diag0_v, 1.0);
+  diag1->compute_diagnostic();
+  auto diag1_v2 = diag1_f.get_view<Real>();
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<>(0, 1),
+      KOKKOS_LAMBDA(const int, bool &local_result) {
+        local_result = (diag1_v2() == diag0_v());
+      },
+      result);
+  REQUIRE(result);
+
+  // other diags
+  auto qc2_v = qc2.get_view<Real **>();
+  Kokkos::deep_copy(qc2_v, 5.0);
+  FieldIdentifier diag2_fid("qc_horiz_avg_manual",
+                            scalar2d_layout.clone().strip_dim(COL), kg / kg,
+                            grid2->name());
+  Field diag2_manual(diag2_fid);
+  diag2_manual.allocate_view();
+  auto diag2_manual_v = diag2_manual.get_view<Real *>();
+  Kokkos::deep_copy(diag2_manual_v, 5.0);
+
+  diag2->set_required_field(qc2);
+  diag2->initialize(t0, RunType::Initial);
+  diag2->compute_diagnostic();
+  auto diag2_f = diag2->get_diagnostic();
+
+  REQUIRE(views_are_equal(diag2_f, diag2_manual));
+
+  auto qc3_v = qc3.get_view<Real ***>();
+  FieldIdentifier diag3_manual_fid("qc_horiz_avg_manual",
+                                   scalar3d_layout.clone().strip_dim(COL),
+                                   kg / kg, grid2->name());
+  Field diag3_manual(diag3_manual_fid);
+  diag3_manual.allocate_view();
+  auto diag3_manual_v = diag3_manual.get_view<Real **>();
+  // calculate diag3_manual by hand
+  TeamPolicy p(dim3 * nlevs, ngcols);
+  Kokkos::parallel_for(
+      "HorizAvgDiag::compute_diagnostic_impl::manual_diag3", p,
+      KOKKOS_LAMBDA(const TeamMember &m) {
+        const int idx = m.league_rank();
+        const int j   = idx / nlevs;
+        const int k   = idx % nlevs;
+        Real sum      = 0.0;
+        Kokkos::parallel_reduce(
+            Kokkos::TeamThreadRange(m, ngcols),
+            [&](const int icol, Real &accum) {
+              accum += (area_v(icol) / atot) * qc3_v(icol, j, k);
+            },
+            sum);
+        Kokkos::single(Kokkos::PerTeam(m),
+                       [&]() { diag3_manual_v(j, k) = sum; });
+      });
+  diag3->set_required_field(qc3);
+  diag3->initialize(t0, RunType::Initial);
+  diag3->compute_diagnostic();
+  auto diag3_f = diag3->get_diagnostic();
+  REQUIRE(views_are_equal(diag3_f, diag3_manual));
+
+  // TODO: add a different flavor of testing
+  // TODO: how to test the MPI part of this rigorously?
+  // TODO: how to test a different type of grid (especially to test the area weighting)
```

Comment:
I don't know if I want to test another type of grid. The only other type is the dyn grid, but the diag would have to be grid-aware, since with dyn grid it would have to slice away 3 dims.